### PR TITLE
Fix IP Pool deletion to correctly remove block affinity

### LINKED ISSUE #519: calicoctl node stop removes enpoint configuration but leaves IPAM data in place
This leads to dangling IP allocations

### LINKED ISSUE #828: No way to clean up an IPAM pool and assignments
This seems to come up fairly frequently on slack:

* User realises they've got an incorrect IPAM pool.
* User adds new pool/removes old pool.
* Stale block assignments hang around, causing trouble.

In particular: routes are advertised for the old blocks and, due to affinity, they get used for assigning new IPs.

Current workaround is to do etcd surgery to remove the blocks.

### PR DESCRIPTION
**Pull request by robbrockbank**
Fixes #828 

Fixes #519

### CODE REVIEW

File: calico_test/tests/st/test_base.py [hunk 1]
```diff
@@ -99,3 +89,28 @@ def assert_ip_connectivity(self, workload_list, ip_pass_list,
...
+            rc = subprocess.check_output(
+                "curl --cacert %s --cert %s --key %s "
+                "-sL https://%s:2379/v2/keys/%s?recursive=%s %s"
+                % (ETCD_CA, ETCD_CERT, ETCD_KEY, ETCD_HOSTNAME_SSL,
+                   path, str(recursive).lower(), " ".join(options)),
+                shell=True)
+        else:
+            rc = subprocess.check_output(
+                "curl -sL http://%s:2379/v2/keys/%s?recursive=%s %s"
+                % (self.ip,
```
Comment (TrimBiggs) [inline calico_test/tests/st/test_base.py]: Seems a little funny to have this alone when some vars below could fit here.

File: calicoctl/calico_ctl/node.py [hunk 2]
```diff
@@ -612,8 +612,16 @@ def node_remove(remove_endpoints):
                         "in the normal way.")
         sys.exit(1)
 
+    # Remove the veths, and release all IPs associated with the endpoint.
+    ips = set()
     for endpoint in endpoints:
         remove_veth(endpoint.name)
+        ips |= {net.ip for net in endpoint.ipv4_nets}
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/node.py]: This isn't a common operator. Can you add a comment saying what this does?

I assume this is setting the `ips` set to include the set of `ips` in both `endpoint.ipv4_nets` and `endpoint.ipv6_nets`, combining all unique IPs into one set to remove.

File: calicoctl/calico_ctl/node.py [hunk 3]
```diff
@@ -612,8 +612,16 @@ def node_remove(remove_endpoints):
...
+        ips |= {net.ip for net in endpoint.ipv6_nets}
+    client.release_ips(ips)
+
+    # Remove the IPAM host data.
+    client.remove_ipam_host(hostname)
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/node.py]: Watch for conflicts with master here, you'll need to update this `hostname` value to `host_to_remove`.

File: calicoctl/calico_ctl/pool.py [hunk 4]
```diff
@@ -188,10 +192,38 @@ def ip_pool_remove(cidrs, version):
...
+            # from the blocks.
+            print "Disabling IP Pool"
+            pool.disabled = True
+            client.set_ip_pool_config(version, pool)
+
+            # Remove affinity from the allocation blocks for the pool.  This
+            # will prevent these blocks from being used for auto-allocations.
+            print "Removing IPAM configuration for pool"
+            client.release_pool_affinities(pool)
+            time.sleep(3)
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/pool.py]: What's this sleep here and the repeated command below?

File: calicoctl/calico_ctl/pool.py [hunk 5]
```diff
@@ -188,10 +192,38 @@ def ip_pool_remove(cidrs, version):
...
+            client.release_pool_affinities(pool)
+
+            print "Deleting IP Pool"
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/pool.py]: I think this is a bit much, just saying the pool is deleted below should be fine.

The `remove_ip_pool` command should be quick and there won't be a long period between this print and the one below, so the "Deleted" print will appear almost immediately after.

File: calicoctl/calico_ctl/container.py [hunk 6]
```diff
@@ -499,14 +498,23 @@ def container_ip_remove(container_id, ip, interface):
...
             ip = result[1][0]
-        pool = get_pool_or_exit(ip)
+
+        # We can't get the pool until we have the IP address.  If we fail to
+        # get the pool we must release the IP.
+        try:
+            pool = get_pool_or_exit(ip)
+        except :
+            client.release_ips({ip})
+            raise
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/container.py]: Test for this try/except?

File: tests/st/no_orchestrator/test_manage_pools.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    """
+    Return the IP address from the output message of calicoctl container add
+    """
+    ip_re = re.compile("IP (.*) added to .*")
+    ip_match = ip_re.match(output.strip())
+    return IPAddress(ip_match.group(1))
+
+class TestManagePools(TestBase):
+    """
+    Test management of IP poola.
```
Comment (TrimBiggs) [inline tests/st/no_orchestrator/test_manage_pools.py]: Poola Party!

File: tests/st/no_orchestrator/test_manage_pools.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            workload3 = host.create_workload("workload3")
+            with self.assertRaises(CalledProcessError):
+                host.calicoctl("container add %s ipv4" % workload3)
+
+            # Add the pool back again, and make sure we get an IP assigned
+            # in the pool range.
+            host.calicoctl("pool add %s" % default_cidr)
+            rc = host.calicoctl("container add %s ipv4" % workload3)
+            ip3 = get_ip_from_container_add_output(rc)
+            assert ip3 in default_cidr
```
Comment (TrimBiggs) [inline tests/st/no_orchestrator/test_manage_pools.py]: Do we want to assert `ip3` is not the same as `ip1` or `ip2`?

File: calicoctl/calico_ctl/pool.py [hunk 9]
```diff
@@ -188,10 +192,38 @@ def ip_pool_remove(cidrs, version):
...
     for cidr in cidrs:
+        # Get the existing IP Pool so that we can disable it,
         try:
-            client.remove_ip_pool(version, IPNetwork(cidr))
+            pool = client.get_ip_pool_config(version, IPNetwork(cidr))
         except KeyError:
             print "%s is not a configured pool." % cidr
+            sys.exit(1)
+
+        try:
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/pool.py]: UT this?

File: calicoctl/calico_ctl/container.py [hunk 10]
```diff
@@ -533,12 +540,13 @@ def get_ip_and_pool(ip):
...
-            return pool
-    print "%s is not found" % cidr
-    sys.exit(1)
+    try:
+        pool = client.get_ip_pool_config(cidr.version, cidr)
+    except KeyError:
+        print "Pool %s is not found" % cidr
+        sys.exit(1)
+    else:
+        return pool
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/container.py]: Guess we didn't have UT for this or below?

File: calicoctl/calico_ctl/container.py [hunk 11]
```diff
@@ -533,12 +540,13 @@ def get_ip_and_pool(ip):
 
 
 def get_pool_by_cidr_or_exit(cidr):
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/container.py]: Mind adding a docstring here?

File: calicoctl/calico_ctl/container.py [hunk 12]
```diff
@@ -499,14 +498,23 @@ def container_ip_remove(container_id, ip, interface):
 
 def get_ip_and_pool(ip):
```
Comment (TrimBiggs) [inline calicoctl/calico_ctl/container.py]: Mind adding a docstring here?

File: calico_test/tests/st/test_base.py [hunk 13]
```diff
@@ -99,3 +89,28 @@ def assert_ip_connectivity(self, workload_list, ip_pass_list,
                 workload.assert_can_ping(ip)
             for ip in ip_fail_list:
                 workload.assert_cant_ping(ip)
+
+    def curl_etcd(self, path, options=[], recursive=True):
```
Comment (TrimBiggs) [inline calico_test/tests/st/test_base.py]: I like this.

**Comment by TrimBiggs** [comment]
Left a bunch of comments.  Let me know if you'd like me to take another look.

File: calicoctl/calico_ctl/node.py (hunk 2 above)
Comment (robbrockbank) [inline calicoctl/calico_ctl/node.py]: Urm, it's a perfectly common operator.  It's completely standard Python.

If it was `blah += [1, 2, 3]` should I comment what the operator is doing?

File: calicoctl/calico_ctl/node.py (hunk 3 above)
Comment (robbrockbank) [inline calicoctl/calico_ctl/node.py]: Yup already rebased and fixed up.

File: calicoctl/calico_ctl/pool.py (hunk 4 above)
Comment (robbrockbank) [inline calicoctl/calico_ctl/pool.py]: Removed first call, beefed up comment.

**Pull request merged**